src/node_worker.cc

```diff
@@ -816,6 +816,100 @@ void Worker::Unref(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
+void Worker::GetHeapStatistics(const FunctionCallbackInfo<Value>& args) {
+  Worker* w;
+  ASSIGN_OR_RETURN_UNWRAP(&w, args.This());
+
+  v8::HeapStatistics heap_stats;
+  w->isolate_->GetHeapStatistics(&heap_stats);
```

Comment:
We should not assume isolate methods are thread safe. To access the worker isolate, please use `w->RequestInterrpt` like in [`Worker::GetHeapSnapshot`](https://github.com/nodejs/node/blob/0a1d5d353bdd18c8beb30991d77ff2925fc5f767/src/node_worker.cc#L896).

Reply:
The flow would be:

1. make the public `getHeapStatistics()` return a promise
2. have a the C++ method take a callback as an argument, which will resolve the promise above.
3. send an interrupt to the other thread and call the callback with the data, using `env->SetImmediateThreadSafe()`.

Is that correct?

Reply:
I think that's about it. Also it's not just about thread safety, the current code is technically violating the V8 requirement to lock the isolate in the thread that attempts to use it and may crash in future versions of V8 where the check is enforced more rigorously.